backend/battling/forms.py

```diff
@@ -8,17 +11,80 @@
     is_team_valid,
     pokemon_in_api,
 )
+from services.email import send_battle_invite
+from users.helper import is_email_valid
 from users.models import User
 
 
+# is_guest = False
+
+
 class CreateBattleForm(forms.ModelForm):
+    opponent = forms.EmailField(
+        required=True,
+    )
+
     class Meta:
         model = Battle
-        fields = ("opponent",)
+        fields = [
+            "creator",
+            "opponent",
+        ]
 
     def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.fields["opponent"].queryset = User.objects.exclude(id=self.initial["creator_id"])
+        super(CreateBattleForm, self).__init__(*args, **kwargs)
+        self.is_guest = False
+        self.fields["creator"].widget = forms.HiddenInput()
+
+    def clean_opponent(self):
+        opponent_email = self.cleaned_data["opponent"]
+        try:
+            opponent = User.objects.get(email=opponent_email)
+        except User.DoesNotExist:
+            self.is_guest = True
+            opponent = User.objects.create(email=opponent_email)
+            random_password = get_random_string(length=64)
+            opponent.set_password(random_password)
+            opponent.save()
+        return opponent
+
+    def clean(self):
+        cleaned_data = super().clean()
+
+        try:
+            opponent = str(cleaned_data["opponent"])
+        except Exception:
+            raise forms.ValidationError("ERROR: Please, type a valid email.")
```

Comment:
Are you sure this error ever happens?

Reply:
Unfortunately... When I tested an email without "@" ou ".", it raised the exception

Reply:
Yeah... Whenever I type an invalid email, it gets this error. In fact, I don't even need the is_email_valid function, because all the errors fall in this Exception anyway

Reply:
But I think the invalid email error happens due to `EmailField`, not this line of code. I don't think this line ever happens.

Reply:
When `clean` is called, all individual clean_<field> have already been called, and all internal validations of the fields like `EmailField` have already been made too.